File: LeetCode Monthly/Mar (2023)/28.py
# https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/description

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)
    
    def strStrRabinKarp(self, haystack: str, needle: str) -> int:
        n = len(needle)
        h = len(haystack)

        hash_n = hash(needle)

        for i in range(h - n + 1):
            if hash(haystack[i: i + n]) == hash_n:
                return i
        return -1




print(Solution().strStrRabinKarp('letcoleede', 'lee'))

File: LeetCode Monthly/Mar (2023)/28.ts
//https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/description

function strStr(haystack: string, needle: string): number {
  const hLen = haystack.length
  const nLen = needle.length
  if (nLen > hLen) return -1
  if (nLen === hLen) return haystack === needle ? 0 : -1

  for (let i = 0; i <= hLen - nLen; i++) {
    if (haystack[i] === needle[0]) {
      if (search(haystack, i, nLen)) return i
    }
  }

  function search(
    haystack: string,
    start: number,
    needleLenght: number
  ): boolean {
    for (let i = 0; i < needleLenght; i++) {
      if (haystack[i + start] !== needle[i]) return false
    }
    return true
  }

  return -1
}

console.log(strStr('dadbutsad', 'sad'))
console.log(strStr('leetcode', 'leeto'))
